src/parser/mod.rs

```diff
@@ -0,0 +1,383 @@
+    let primary = |pair: Pair<Rule>| parse_logical_expression(pair);
+
+    let infix = |lhs: Result<Expression>, op: Pair<Rule>, rhs: Result<Expression>| {
+        let result = match op.as_rule() {
+            Rule::logical_or => Expression::new(ExpressionValue::Logical(LogicalExpression::new(
+                lhs?,
+                rhs?,
+                LogicalOperator::Or,
+            ))),
+            Rule::logical_and => Expression::new(ExpressionValue::Logical(LogicalExpression::new(
+                lhs?,
+                rhs?,
+                LogicalOperator::And,
+            ))),
+            _ => bail!("parse_logical_expression(infix): invalid grammar? {:?}", op.as_rule()),
+        };
+        Ok(result)
+    };
+
+    match pair.as_rule() {
+        Rule::logical_value => parse_logical_value(pair),
+        Rule::logical_expression => LOGICAL_CLIMBER.climb(pair.into_inner(), primary, infix),
+        _ => bail!("parse_logical_expression: invalid grammar? {:?}", pair.as_rule()),
+    }
+}
+
+//
+// basic_expression_filter = { basic_expression ~ filter* }
+//
+fn parse_basic_expression_with_filters(pair: Pair<Rule>) -> Result<Expression> {
+    let mut expression = None;
+    let mut filters = vec![];
+
+    for p in pair.into_inner() {
+        match p.as_rule() {
+            Rule::basic_expression => expression = Some(parse_basic_expression(p)?),
+            Rule::filter => filters.push(parse_filter(p)?),
+            _ => bail!("parse_basic_expression_with_filters: invalid grammar? {:?}", p),
+        };
+    }
+
+    let exp = expression
+        .ok_or_else(|| "parse_basic_expression_with_filters: invalid grammar, unable to create expression")?;
+    Ok(Expression::new_with_filters(exp, filters))
+}
+
+//
+// double_quoted_string  = @{ "\"" ~ (!("\"") ~ ANY)* ~ "\""}
+// single_quoted_string  = @{ "\'" ~ (!("\'") ~ ANY)* ~ "\'"}
+// backquoted_quoted_string  = @{ "`" ~ (!("`") ~ ANY)* ~ "`"}
+//
+// string = @{ double_quoted_string | single_quoted_string | backquoted_quoted_string }
+//
+fn remove_string_quotes(input: &str) -> Result<String> {
+    let result = match input
+        .chars()
+        .next()
+        .ok_or_else(|| "remove_string_quotes: invalid grammar, unable to remove quotes")?
+    {
+        '"' => input.replace('"', "").to_string(),
+        '\'' => input.replace('\'', "").to_string(),
+        '`' => input.replace('`', "").to_string(),
+        _ => bail!("remove_string_quotes: invalid grammar? {}", input),
+    };
+    Ok(result)
+}
+
+// all_chars = _{'a'..'z' | 'A'..'Z' | "_" | '0'..'9'}
+// identifier = @{
+//     ('a'..'z' | 'A'..'Z' | "_") ~
+//     all_chars*
+// }
+//
+// square_brackets = _{
+//     "[" ~ (integer | string | dotted_square_bracket_identifier) ~ "]"
+// }
+//
+// dotted_square_bracket_identifier = ${
+//     identifier ~ ( ("." ~ identifier) | square_brackets )*
+// }
+//
+fn _parse_dotted_square_bracket_identifier(pair: Pair<Rule>) -> Result<Identifier> {
```

Comment:
It may no be immediately obvious why the function name starts with underscore - I was thinking for few seconds that this has to do with the `unused` convention in Rust. My current understanding is that the result is wrapped in the function `parse_dotted_square_bracket_identifier` (without the underscore in the name) but also this one calls itself recursively - which makes this difficult to name, I agree. Just leaving a note here without much suggestion tbh - maybe a name change could help, maybe a sentence in the function documentation ?

Reply:
Renamed, not sure it helps a lot. Anyway, I'll probably rename all of these functions (include expression in the name for example to be more clear what it parses).